// @atlaskit/tokens/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ThemeMutationObserver", {
  enumerable: true,
  get: function get() {
    return _themeMutationObserver.ThemeMutationObserver;
  }
});
Object.defineProperty(exports, "getGlobalTheme", {
  enumerable: true,
  get: function get() {
    return _getGlobalTheme.getGlobalTheme;
  }
});
Object.defineProperty(exports, "getSSRAutoScript", {
  enumerable: true,
  get: function get() {
    return _setGlobalTheme.getSSRAutoScript;
  }
});
Object.defineProperty(exports, "getThemeHtmlAttrs", {
  enumerable: true,
  get: function get() {
    return _setGlobalTheme.getThemeHtmlAttrs;
  }
});
Object.defineProperty(exports, "getThemeStyles", {
  enumerable: true,
  get: function get() {
    return _setGlobalTheme.getThemeStyles;
  }
});
Object.defineProperty(exports, "getTokenValue", {
  enumerable: true,
  get: function get() {
    return _getTokenValue.default;
  }
});
Object.defineProperty(exports, "setGlobalTheme", {
  enumerable: true,
  get: function get() {
    return _setGlobalTheme.default;
  }
});
Object.defineProperty(exports, "themeConfig", {
  enumerable: true,
  get: function get() {
    return _themeConfig.default;
  }
});
Object.defineProperty(exports, "themeObjectToString", {
  enumerable: true,
  get: function get() {
    return _themeStateTransformer.themeObjectToString;
  }
});
Object.defineProperty(exports, "themeStringToObject", {
  enumerable: true,
  get: function get() {
    return _themeStateTransformer.themeStringToObject;
  }
});
Object.defineProperty(exports, "token", {
  enumerable: true,
  get: function get() {
    return _getToken.default;
  }
});
Object.defineProperty(exports, "useThemeObserver", {
  enumerable: true,
  get: function get() {
    return _useThemeObserver.useThemeObserver;
  }
});
var _getToken = _interopRequireDefault(require("./get-token"));
var _getTokenValue = _interopRequireDefault(require("./get-token-value"));
var _setGlobalTheme = _interopRequireWildcard(require("./set-global-theme"));
var _themeConfig = _interopRequireDefault(require("./theme-config"));
var _useThemeObserver = require("./use-theme-observer");
var _themeMutationObserver = require("./theme-mutation-observer");
var _getGlobalTheme = require("./get-global-theme");
var _themeStateTransformer = require("./utils/theme-state-transformer");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
