// @atlaskit/textfield/dist/cjs/component-tokens.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.transparent = exports.textColor = exports.subtleBorderColorHover = exports.placeholderTextColor = exports.disabledTextColor = exports.disabledBackgroundColor = exports.defaultBorderColorFocus = exports.defaultBorderColor = exports.defaultBackgroundColorHover = exports.defaultBackgroundColorFocus = exports.defaultBackgroundColor = void 0;
var _colors = require("@atlaskit/theme/colors");
var disabledBackgroundColor = {
  light: "var(--ds-background-disabled, ".concat(_colors.N10, ")"),
  dark: "var(--ds-background-disabled, ".concat(_colors.DN10, ")")
};
exports.disabledBackgroundColor = disabledBackgroundColor;
var defaultBackgroundColor = {
  light: "var(--ds-background-input, ".concat(_colors.N10, ")"),
  dark: "var(--ds-background-input, ".concat(_colors.DN10, ")")
};
exports.defaultBackgroundColor = defaultBackgroundColor;
var defaultBackgroundColorFocus = {
  light: "var(--ds-background-input-pressed, ".concat(_colors.N0, ")"),
  dark: "var(--ds-background-input-pressed, ".concat(_colors.DN10, ")")
};
exports.defaultBackgroundColorFocus = defaultBackgroundColorFocus;
var defaultBackgroundColorHover = {
  light: "var(--ds-background-input-hovered, ".concat(_colors.N30, ")"),
  dark: "var(--ds-background-input-hovered, ".concat(_colors.DN30, ")")
};
exports.defaultBackgroundColorHover = defaultBackgroundColorHover;
var defaultBorderColor = {
  light: "var(--ds-border-input, ".concat(_colors.N40, ")"),
  dark: "var(--ds-border-input, ".concat(_colors.DN40, ")")
};
exports.defaultBorderColor = defaultBorderColor;
var defaultBorderColorFocus = {
  light: "var(--ds-border-focused, ".concat(_colors.B100, ")"),
  dark: "var(--ds-border-focused, ".concat(_colors.B75, ")")
};
exports.defaultBorderColorFocus = defaultBorderColorFocus;
var subtleBorderColorHover = {
  light: "var(--ds-border-input, transparent)",
  dark: "var(--ds-border-input, transparent)"
};
exports.subtleBorderColorHover = subtleBorderColorHover;
var transparent = {
  light: 'transparent',
  dark: 'transparent'
};
exports.transparent = transparent;
var textColor = {
  light: "var(--ds-text, ".concat(_colors.N900, ")"),
  dark: "var(--ds-text, ".concat(_colors.DN600, ")")
};
exports.textColor = textColor;
var disabledTextColor = {
  light: "var(--ds-text-disabled, ".concat(_colors.N70, ")"),
  dark: "var(--ds-text-disabled, ".concat(_colors.DN90, ")")
};
exports.disabledTextColor = disabledTextColor;
var placeholderTextColor = {
  light: "var(--ds-text-subtlest, ".concat(_colors.N100, ")"),
  dark: "var(--ds-text-subtlest, ".concat(_colors.DN90, ")")
};
exports.placeholderTextColor = placeholderTextColor;
