// @atlaskit/tokens/dist/cjs/tokens/atlassian-dark/color/text.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var color = {
  color: {
    text: {
      '[default]': {
        value: 'DarkNeutral1000'
      },
      subtle: {
        value: 'DarkNeutral800'
      },
      subtlest: {
        value: 'DarkNeutral700'
      },
      inverse: {
        value: 'DarkNeutral0'
      },
      disabled: {
        value: 'DarkNeutral400A'
      },
      brand: {
        value: 'Blue400'
      },
      selected: {
        value: 'Blue400'
      },
      danger: {
        value: 'Red300'
      },
      warning: {
        '[default]': {
          value: 'Yellow300'
        },
        inverse: {
          value: 'DarkNeutral0'
        }
      },
      success: {
        value: 'Green300'
      },
      information: {
        value: 'Blue300'
      },
      discovery: {
        value: 'Purple300'
      }
    },
    link: {
      '[default]': {
        value: 'Blue400'
      },
      pressed: {
        value: 'Blue300'
      }
    }
  }
};
var _default = color;
exports.default = _default;
