// @atlaskit/tokens/dist/cjs/artifacts/theme-import-map.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || (0, _typeof2.default)(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * THIS FILE WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 *
 * This file contains a dynamic import for each theme this package exports.
 * Themes are loaded asynchronously at runtime to minimise the amount of CSS we send to the client.
 * This allows users to compose their themes and only use the tokens that are requested.
 * When a new theme is created, the import should automatically be added to the map
 *
 * @codegen <<SignedSource::3466f435e0d35c0443ddeafe9240dbc2>>
 * @codegenCommand yarn build tokens
 */

var themeImportsMap = {
  'light': function light() {
    return Promise.resolve().then(function () {
      return _interopRequireWildcard(require('./themes/atlassian-light'));
    });
  },
  'dark': function dark() {
    return Promise.resolve().then(function () {
      return _interopRequireWildcard(require('./themes/atlassian-dark'));
    });
  },
  'legacy-light': function legacyLight() {
    return Promise.resolve().then(function () {
      return _interopRequireWildcard(require('./themes/atlassian-legacy-light'));
    });
  },
  'legacy-dark': function legacyDark() {
    return Promise.resolve().then(function () {
      return _interopRequireWildcard(require('./themes/atlassian-legacy-dark'));
    });
  },
  'spacing': function spacing() {
    return Promise.resolve().then(function () {
      return _interopRequireWildcard(require('./themes/atlassian-spacing'));
    });
  },
  'typography': function typography() {
    return Promise.resolve().then(function () {
      return _interopRequireWildcard(require('./themes/atlassian-typography'));
    });
  },
  'shape': function shape() {
    return Promise.resolve().then(function () {
      return _interopRequireWildcard(require('./themes/atlassian-shape'));
    });
  },
  'dark-iteration': function darkIteration() {
    return Promise.resolve().then(function () {
      return _interopRequireWildcard(require('./themes/atlassian-dark-iteration'));
    });
  }
};
var _default = themeImportsMap;
exports.default = _default;
