// @atlaskit/tokens/dist/esm/palettes/shape-palette.js
/**
 * Types are inferred from the base tokens below
 */

var baseSizeTokens = {
  Size0: {
    value: '0',
    attributes: {
      group: 'shape'
    }
  },
  Size050: {
    value: '0.0625rem',
    attributes: {
      group: 'shape'
    }
  },
  Size100: {
    value: '0.125rem',
    attributes: {
      group: 'shape'
    }
  }
};
var baseRadiusTokens = {
  Radius050: {
    value: '0.125rem',
    attributes: {
      group: 'shape'
    }
  },
  Radius100: {
    value: '0.25rem',
    attributes: {
      group: 'shape'
    }
  },
  Radius200: {
    value: '0.5rem',
    attributes: {
      group: 'shape'
    }
  },
  Radius300: {
    value: '0.75rem',
    attributes: {
      group: 'shape'
    }
  },
  Radius400: {
    value: '1rem',
    attributes: {
      group: 'shape'
    }
  },
  RadiusCircle: {
    value: '50%',
    attributes: {
      group: 'shape'
    }
  }
};
var shapePalette = {
  border: {
    radius: baseRadiusTokens,
    width: baseSizeTokens
  }
};
export default shapePalette;
