// @atlaskit/tokens/dist/cjs/tokens/default/typography/font-weight.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var font = {
  font: {
    weight: {
      regular: {
        attributes: {
          group: 'fontWeight',
          state: 'active',
          introduced: '0.10.33',
          description: 'Helpful guidance goes here'
        }
      },
      medium: {
        attributes: {
          group: 'fontWeight',
          state: 'active',
          introduced: '0.10.33',
          description: 'Helpful guidance goes here'
        }
      },
      semibold: {
        attributes: {
          group: 'fontWeight',
          state: 'active',
          introduced: '0.10.33',
          description: 'Helpful guidance goes here'
        }
      },
      bold: {
        attributes: {
          group: 'fontWeight',
          state: 'active',
          introduced: '0.10.33',
          description: 'Helpful guidance goes here'
        }
      }
    }
  }
};
var _default = font;
exports.default = _default;
